lib/update-buffer/src/lib.rs

```diff
@@ -0,0 +1,313 @@
+            panic!();
+        }
+
+        // Lock ordering: We acquire the owner lock first.
+        let mut owner = self.current_owner.lock();
+        if let Some(owner) = owner.as_ref() {
+            // If `owner` is `Some(_)`, we know there is an outstanding
+            // `UpdateBuffer`; return an error.
+            Err(owner.clone())
+        } else {
+            // `owner` is `None`, which means either:
+            //
+            // 1. No `UpdateBuffer` exists (this lock will succeed immediately)
+            // 2. An `UpdateBuffer` is in the process of being dropped (this
+            //    lock will succeed very soon)
+            //
+            // Either way, we must first acquire the data lock _before_
+            // unlocking `owner`, to avoid TOCTOU races with concurrent calls to
+            // this method.
+            //
+            // This has the potential to deadlock with `UpdateBuffer::drop()`,
+            // which acquires the `owner` lock while still holding the `data`
+            // lock (the reverse order from this method). Normally acquiring
+            // locks in opposite order is a good way to deadlock, but we avoid
+            // it in this case due to our use of `owner`: We only attempt to
+            // acquire the `data` lock if `owner` is `None`. If `owner` is
+            // `Some(_)`, we immediately release the lock and return. If there
+            // is an outstanding `UpdateBuffer`, `owner` remains `Some(_)` until
+            // its drop method is able to acquire the `owner` lock (which it
+            // will always be able to, because we only hold it momentarily if
+            // it's `Some(_)`) and set it back to `None`, before the drop
+            // continues and `data` is unlocked.
+            let data = self.data.lock();
+
+            // We now own the data; record our `new_owner` tag and drop the
+            // owner lock. Any other caller to this method will see our owner
+            // until the `UpdateBuffer` we're about to create is dropped, since
+            // `owner` will remain `Some(_)` until that drop impl runs.
+            *owner = Some(new_owner.clone());
+            mem::drop(owner);
+
+            Ok(UpdateBuffer {
+                owner: &self.current_owner,
+                data,
+                len: 0,
+                cap: capacity,
+            })
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct UpdateBuffer<'a, T, const N: usize> {
+    owner: &'a Mutex<Option<T>>,
+    data: MutexGuard<'a, [u8; N]>,
+    len: usize,
+    cap: usize,
+}
+
+impl<T, const N: usize> Drop for UpdateBuffer<'_, T, N> {
+    fn drop(&mut self) {
+        // Avoiding deadlocks: We currently own the `data` lock and are
+        // attempting to acquire the `owner` lock. This is the reverse order
+        // from `BufferMutex::borrow()` above. See the comment in it above for
+        // the reasoning why this will not deadlock.
+        *self.owner.lock() = None;
```

Comment:
Is it possible to explicitly drop `self.data` here?

That would make the ordering more obvious, but not sure if it's allowed.

Reply:
It is not possible, no. At least not without wrapping `self.data` in an `Option` (or some moral equivalent), which would allow us to `.take()` it and explicitly drop it.

After rereading, I could change `borrow()` to not lock the data until after it unlocks the owner, which would eliminate basically all the comments regarding inverted lock ordering (since the only lock-within-a-lock would be right here: `borrow()` would acquire owner, release it, then acquire data).